### PR TITLE
KAFKA-10520; Ensure transactional producers poll if leastLoadedNode not available with max.in.flight=1

### PR DESCRIPTION
**Pull request by rajinisivaram**
We currently stop polling in `Sender` in a transactional producer if there is only one broker in the bootstrap server list and `max.in.flight.requests.per.connection=1` and Metadata response is pending when InitProducerId request is ready to be sent. In this scenario, we attempt to send FindCoordinator to `leastLoadedNode`, but since that is blocked due to `max.in.flight=1` as a result of the pending metadata response, we never unblock unless we poll. This PR ensures we poll in this case.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java [hunk 1]
```diff
@@ -444,10 +444,20 @@ private boolean maybeSendAndPollTransactionalRequest() {
...
+            if (targetNode != null) {
+                awaitNodeReady(targetNode, coordinatorType);
+            } else if (coordinatorType != null) {
+                log.trace("Coordinator not known for {}, will retry {} after finding coordinator.", coordinatorType, requestBuilder.apiKey());
                 maybeFindCoordinatorAndRetry(nextRequestHandler);
                 return true;
+            } else {
+                log.trace("No nodes available to send requests, polling until a node is ready.");
+                client.poll(retryBackoffMs, time.milliseconds());
+                return true;
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: Don't we loose the current `nextRequestHandler` when we end up in this branch?

For instance, if `nextRequestHandler` is a `FindCoordinatorHandler` and there a no nodes available, `targetNode` is `null` and `coordinatorType` is `null` as well so we end up here and poll. We don't do anything with `nextRequestHandler` and return so it is gone. I suppose that it is not an issue for `FindCoordinatorHandler` as a new one will be enqueued automatically when another `TxnRequestHandler` handler is processed and the coordinator is unknown.

We may want to push back `nextRequestHandler` to the queue with `transactionManager.retry(nextRequestHandler)` in oder to handle all the cases.

I am not sure if that could really happen with any other `TxnRequestHandler` type though. What do you think?

Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: @dajac Thanks for the review. As far as I can tell, `nextRequestHandler` returned by `transactionManager.nextRequest` would have retained the request in this case since it does a peek and removes the element only under certain conditions. I was trying to limit the amount of change to the minimal necessary to handle max.in.flight=1, do you think we should do more to look into the other cases?

Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: @rajinisivaram It seems that a non-null `nextRequestHandler` would have been removed from the queue: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java#L872.

I've got the impression that `InitProducerIdHandler` may hit that code path as well as it could have a null `coordinatorType` when the transaction id is not set (idempotent producer). I think that we could unit test this actually by reusing your `testInitProducerIdWithMaxInFlightOne` but without setting the transactional id. Should we add this test to be on the safe side?

Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: @dajac Sorry, that was my mistake. Updated code and added test.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java [hunk 2]
```diff
@@ -512,21 +523,15 @@ public boolean isRunning() {
         return running;
     }
 
-    private Node awaitNodeReady(FindCoordinatorRequest.CoordinatorType coordinatorType) throws IOException {
-        Node node = coordinatorType != null ?
-                transactionManager.coordinator(coordinatorType) :
-                client.leastLoadedNode(time.milliseconds());
-
+    private void awaitNodeReady(Node node, FindCoordinatorRequest.CoordinatorType coordinatorType) throws IOException {
         if (node != null && NetworkClientUtils.awaitReady(client, node, time, requestTimeoutMs)) {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: ```node != null``` Is this check necessary?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java [hunk 3]
```diff
@@ -444,10 +444,21 @@ private boolean maybeSendAndPollTransactionalRequest() {
...
         Node targetNode = null;
         try {
-            targetNode = awaitNodeReady(nextRequestHandler.coordinatorType());
-            if (targetNode == null) {
+            FindCoordinatorRequest.CoordinatorType coordinatorType = nextRequestHandler.coordinatorType();
+            targetNode = coordinatorType != null ?
+                    transactionManager.coordinator(coordinatorType) :
+                    client.leastLoadedNode(time.milliseconds());
+            if (targetNode != null) {
+                awaitNodeReady(targetNode, coordinatorType);
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: How to deal with the non-ready targetNode after calling ```awaitNodeReady``` ?

File: core/src/test/scala/integration/kafka/api/TransactionsWithMaxInFlightOneTest.scala [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+import scala.collection.Seq
+import scala.collection.mutable.Buffer
+import scala.jdk.CollectionConverters._
+
+/**
+ * This is used to test transactions with one broker and `max.in.flight.requests.per.connection=1`.
+ * A single broker is used to verify edge cases where different requests are queued on the same connection.
+ */
+class TransactionsWithMaxInFlightOneTest extends KafkaServerTestHarness {
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/TransactionsWithMaxInFlightOneTest.scala:39]: this new test passes on my local even though I revert the change of ```Sender```.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java (hunk 2 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: No, we don't need it anymore, removed.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java (hunk 3 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: @chia7712 Thanks for the review. Good point, added check to retry in that case, also added unit test.

File: core/src/test/scala/integration/kafka/api/TransactionsWithMaxInFlightOneTest.scala (hunk 4 above)
Comment (rajinisivaram) [reply core/src/test/scala/integration/kafka/api/TransactionsWithMaxInFlightOneTest.scala:39]: Yes, the issue this PR fixes is a timing issue that the new unit tests catch, but it is unlikely that we hit it in integration tests. I added this integration test anyway to make sure the fix hasn't broken anything. I was in two minds about whether to check it in, but included anyway since I thought it may help catch regressions with max.in.flight=1 in future. Also, Jenkins builds on slower machines may occasionally hit timing windows we don't usually see otherwise. Don't mind removing if we think it is not worth the maintenance cost.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java (hunk 1 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java]: Thanks. I will make another pass now.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java [hunk 5]
```diff
@@ -444,10 +444,25 @@ private boolean maybeSendAndPollTransactionalRequest() {
...
+                }
+            } else if (coordinatorType != null) {
+                log.trace("Coordinator not known for {}, will retry {} after finding coordinator.", coordinatorType, requestBuilder.apiKey());
                 maybeFindCoordinatorAndRetry(nextRequestHandler);
                 return true;
+            } else {
+                log.trace("No nodes available to send requests, will poll and retry when until a node is ready.");
+                transactionManager.retry(nextRequestHandler);
+                client.poll(retryBackoffMs, time.milliseconds());
+                return true;
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:465]: Looking at this branch again, sorry :). I was comparing it with the previous behavior and I have noticed that we would request a refresh of the metadata when the same conditions were met. That happened here:
```
    private void maybeFindCoordinatorAndRetry(TransactionManager.TxnRequestHandler nextRequestHandler) {
        if (nextRequestHandler.needsCoordinator()) {
            transactionManager.lookupCoordinator(nextRequestHandler);
        } else {
            // For non-coordinator requests, sleep here to prevent a tight loop when no node is available
            time.sleep(retryBackoffMs);
            metadata.requestUpdate();
        }

        transactionManager.retry(nextRequestHandler);
    }
```
When no node is available and `coordinatorType != null`, we ended up in the else branch here. I wonder if not doing `metadata.requestUpdate()` in our new implementation could be problematic and I also wonder if we could just swap that `time.sleep(retryBackoffMs)` by `client.poll(retryBackoffMs, time.milliseconds())` to achieve the same goal. The difference is that another metadata request would be sent in our particular case.

Have you noticed this small difference?

Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:465]: @dajac Yes, I was thinking about this earlier when adding the retry. We get to this `else` path when leastLoadedNode is null. For the other cases, we would have taken one of the other paths, which retain existing behaviour (unless I missed something, again!) When least loaded node is null, the old behaviour didn't quite work because you have to poll to change that state. And you don't have anywhere to send metadata requests to. So just polling seemed to be sufficient for this case. What do you think?

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:465]: I agree polling seems sufficient. We will still have an opportunity to refresh metadata if the current connection fails for some reason.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java [hunk 6]
```diff
@@ -2667,4 +2760,43 @@ private void assertFutureFailure(Future<?> future, Class<? extends Exception> ex
         }
     }
 
+    private void createMockClientWithMaxFlightOneMetadataPending() {
+        client = new MockClient(time, metadata) {
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:2764]: Wonder if we should consider adding max inflight behavior directly to `MockClient`. Seems like a notable difference from `NetworkClient`.

**Review by hachikuji — APPROVED** [review]
Thanks for the fix. LGTM.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java (hunk 5 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:465]: @rajinisivaram Yeah, I do agree. Polling seems sufficient in this case. Thanks for the clarification.

**Review by dajac — APPROVED** [review]
LGTM. Thanks for the fix.

**Comment by hachikuji** [comment]
Merging to trunk and 2.7.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java (hunk 6 above)
Comment (rajinisivaram) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:2764]: @hachikuji Thanks for reviewing and merging! I have opened https://issues.apache.org/jira/browse/KAFKA-10626 to add max.in.flight to MockClient.
